File: internal/ast/value_type.go
package ast

import (
	"fmt"
	"strings"
)

type ValueType uint8

const (
	ValueTypeUnknown ValueType = iota

	ValueTypeDoc
	ValueTypeSlice
	ValueTypeArray
	ValueTypeObject
	ValueTypeMap

	ValueTypeString
	ValueTypeBytes
	ValueTypeBool

	ValueTypeInt
	ValueTypeInt8
	ValueTypeInt16
	ValueTypeInt32
	ValueTypeInt64
	ValueTypeUint
	ValueTypeUint8
	ValueTypeUint16
	ValueTypeUint32
	ValueTypeUint64

	ValueTypeFloat32
	ValueTypeFloat64

	ValueTypeBigInt
	ValueTypeDateTime
	ValueTypeDate
	ValueTypeTime

	ValueTypeUUID
	ValueTypeDecimal
	ValueTypeIP
	ValueTypeURL
	ValueTypeEmail

	ValueTypeEnum

	ValueTypeImage
	ValueTypeVideo

	vtUnknownStr = "unknown"

	vtDocStr    = "doc"
	vtArrayStr  = "arr"
	vtSliceStr  = "slice"
	vtObjectStr = "obj"
	vtMapStr    = "map"

	vtStringStr = "str"
	vtBytesStr  = "bytes"
	vtBoolStr   = "bool"

	vtIStr   = "i"
	vtI8Str  = "i8"
	vtI16Str = "i16"
	vtI32Str = "i32"
	vtI64Str = "i64"
	vtUStr   = "u"
	vtU8Str  = "u8"
	vtU16Str = "u16"
	vtU32Str = "u32"
	vtU64Str = "u64"

	vtF32Str = "f32"
	vtF64Str = "f64"

	vtBiStr       = "bi"
	vtDateTimeStr = "datetime"
	vtDateStr     = "date"
	vtTimeStr     = "time"
	vtUUIDStr     = "uuid"
	vtDecimalStr  = "decimal"
	vtIPStr       = "ip"
	vtURLStr      = "url"
	vtEmailStr    = "email"

	vtEnumStr = "enum"

	vtImageStr = "image"
	vtVideoStr = "video"
)

func (v ValueType) String() string {
	switch v {
	case ValueTypeUnknown:
		return vtUnknownStr
	case ValueTypeDoc:
		return vtDocStr
	case ValueTypeArray:
		return vtArrayStr
	case ValueTypeSlice:
		return vtSliceStr
	case ValueTypeObject:
		return vtObjectStr
	case ValueTypeMap:
		return vtMapStr
	case ValueTypeString:
		return vtStringStr
	case ValueTypeBytes:
		return vtBytesStr
	case ValueTypeBool:
		return vtBoolStr
	case ValueTypeInt:
		return vtIStr
	case ValueTypeInt8:
		return vtI8Str
	case ValueTypeInt16:
		return vtI16Str
	case ValueTypeInt32:
		return vtI32Str
	case ValueTypeInt64:
		return vtI64Str
	case ValueTypeUint:
		return vtUStr
	case ValueTypeUint8:
		return vtU8Str
	case ValueTypeUint16:
		return vtU16Str
	case ValueTypeUint32:
		return vtU32Str
	case ValueTypeUint64:
		return vtU64Str
	case ValueTypeFloat32:
		return vtF32Str
	case ValueTypeFloat64:
		return vtF64Str
	case ValueTypeBigInt:
		return vtBiStr
	case ValueTypeDateTime:
		return vtDateTimeStr
	case ValueTypeDate:
		return vtDateStr
	case ValueTypeTime:
		return vtTimeStr
	case ValueTypeUUID:
		return vtUUIDStr
	case ValueTypeDecimal:
		return vtDecimalStr
	case ValueTypeIP:
		return vtIPStr
	case ValueTypeURL:
		return vtURLStr
	case ValueTypeEmail:
		return vtEmailStr
	case ValueTypeEnum:
		return vtEnumStr
	case ValueTypeImage:
		return vtImageStr
	case ValueTypeVideo:
		return vtVideoStr
	default:
		return fmt.Sprintf("ValueType(%d)", v)
	}
}

var strToValueType = map[string]ValueType{
	vtUnknownStr:  ValueTypeUnknown,
	vtDocStr:      ValueTypeDoc,
	vtArrayStr:    ValueTypeArray,
	vtSliceStr:    ValueTypeSlice,
	vtObjectStr:   ValueTypeObject,
	vtMapStr:      ValueTypeMap,
	vtStringStr:   ValueTypeString,
	vtBytesStr:    ValueTypeBytes,
	vtBoolStr:     ValueTypeBool,
	vtIStr:        ValueTypeInt,
	vtI8Str:       ValueTypeInt8,
	vtI16Str:      ValueTypeInt16,
	vtI32Str:      ValueTypeInt32,
	vtI64Str:      ValueTypeInt64,
	vtUStr:        ValueTypeUint,
	vtU8Str:       ValueTypeUint8,
	vtU16Str:      ValueTypeUint16,
	vtU32Str:      ValueTypeUint32,
	vtU64Str:      ValueTypeUint64,
	vtF32Str:      ValueTypeFloat32,
	vtF64Str:      ValueTypeFloat64,
	vtBiStr:       ValueTypeBigInt,
	vtDateTimeStr: ValueTypeDateTime,
	vtDateStr:     ValueTypeDate,
	vtTimeStr:     ValueTypeTime,
	vtUUIDStr:     ValueTypeUUID,
	vtDecimalStr:  ValueTypeDecimal,
	vtIPStr:       ValueTypeIP,
	vtURLStr:      ValueTypeURL,
	vtEmailStr:    ValueTypeEmail,
	vtEnumStr:     ValueTypeEnum,
	vtImageStr:    ValueTypeImage,
	vtVideoStr:    ValueTypeVideo,
}

func ParseValueType(s string) (ValueType, error) {
	s = strings.ToLower(s)

	vt, ok := strToValueType[s]
	if !ok {
		return ValueTypeUnknown, fmt.Errorf("Invalid ValueType string: %s", s)
	}
	return vt, nil
}

File: mm-py/metamessage/ir/value_type.py
"""
ValueType enum for MetaMessage.
"""
from enum import IntEnum


class ValueType(IntEnum):
    Unknown = 0
    Doc = 1
    Vec = 2
    Arr = 3
    Obj = 4
    Map = 5
    Str = 6
    Bytes = 7
    Bool = 8
    I = 9
    I8 = 10
    I16 = 11
    I32 = 12
    I64 = 13
    U = 14
    U8 = 15
    U16 = 16
    U32 = 17
    U64 = 18
    F32 = 19
    F64 = 20
    Bigint = 21
    Datetime = 22
    Date = 23
    Time = 24
    Uuid = 25
    Decimal = 26
    Ip = 27
    Url = 28
    Email = 29
    Enums = 30
    Media = 31

    def __str__(self) -> str:
        mapping = {
            ValueType.Unknown: "unknown",
            ValueType.Doc: "doc",
            ValueType.Arr: "arr",
            ValueType.Vec: "vec",
            ValueType.Obj: "obj",
            ValueType.Map: "map",
            ValueType.Str: "str",
            ValueType.Bytes: "bytes",
            ValueType.Bool: "bool",
            ValueType.I: "i",
            ValueType.I8: "i8",
            ValueType.I16: "i16",
            ValueType.I32: "i32",
            ValueType.I64: "i64",
            ValueType.U: "u",
            ValueType.U8: "u8",
            ValueType.U16: "u16",
            ValueType.U32: "u32",
            ValueType.U64: "u64",
            ValueType.F32: "f32",
            ValueType.F64: "f64",
            ValueType.Bigint: "bigint",
            ValueType.Datetime: "datetime",
            ValueType.Date: "date",
            ValueType.Time: "time",
            ValueType.Uuid: "uuid",
            ValueType.Decimal: "decimal",
            ValueType.Ip: "ip",
            ValueType.Url: "url",
            ValueType.Email: "email",
            ValueType.Enums: "enums",
            ValueType.Media: "media",
        }
        return mapping.get(self, "ValueType(%d)" % self.value)


_str_to_value_type = {
    "unknown": ValueType.Unknown, "doc": ValueType.Doc, "vec": ValueType.Vec,
    "arr": ValueType.Arr, "obj": ValueType.Obj, "map": ValueType.Map,
    "str": ValueType.Str, "bytes": ValueType.Bytes, "bool": ValueType.Bool,
    "i": ValueType.I, "i8": ValueType.I8, "i16": ValueType.I16, "i32": ValueType.I32, "i64": ValueType.I64,
    "u": ValueType.U, "u8": ValueType.U8, "u16": ValueType.U16, "u32": ValueType.U32, "u64": ValueType.U64,
    "f32": ValueType.F32, "f64": ValueType.F64,
    "bigint": ValueType.Bigint, "datetime": ValueType.Datetime, "date": ValueType.Date, "time": ValueType.Time,
    "uuid": ValueType.Uuid, "decimal": ValueType.Decimal, "ip": ValueType.Ip, "url": ValueType.Url,
    "email": ValueType.Email, "enums": ValueType.Enums,
    "media": ValueType.Media,
}


def parse_value_type(s: str) -> ValueType:
    s = s.lower().strip()
    if s in _str_to_value_type:
        return _str_to_value_type[s]
    raise ValueError("Invalid ValueType string: %s" % s)